autoscale/src/main/java/com/sequenceiq/periscope/service/AltusMachineUserService.java

```diff
@@ -97,4 +108,9 @@ private void syncEnvironment(String accountId, String machineUserCrn, String env
         LOGGER.info("Finished invoking freeIpa user Sync with operation: {}, for environmentCrn: {} and machineUserCrn: {}", statusResponse.getOperationId(),
                 environmentCrn, machineUserCrn);
     }
+
+    private boolean userSyncNeeded(String stackCrn) {
+        String version = cloudbreakDetailsService.getCloudbreakDetailsByStackCrn(stackCrn).getVersion();
```

Comment:
This likely needs a null check. The field isn't guaranteed to be set. 
See: https://github.com/hortonworks/cloudbreak/blob/master/core/src/main/java/com/sequenceiq/cloudbreak/converter/v4/stacks/StackToStackV4ResponseConverter.java#L226

Also: Can you please explore what the various states in ComponentType mean ... specifically SALT_STATE looks interesting.

Also for reference: https://github.com/hortonworks/cloudbreak/blob/master/core/src/main/java/com/sequenceiq/cloudbreak/service/stack/StackStopRestrictionService.java#L72

If the sale version is in fact being tracked ... should be possible to add this information into StackV4Response, similar to the CB details. This is only if this tracks the current version of the sale scripts (assuming it tracks salt scripts), rather than the version at deployment.

Reply:
Looked into ComponentType.SALT_STATE, the salt component is created and updated in ClusterBootstrapper -> https://github.com/hortonworks/cloudbreak/blob/6336fda3fe3c7393d68637a838f045dab9f0590f/core/src/main/java/com/sequenceiq/cloudbreak/core/bootstrap/service/ClusterBootstrapper.java#L254-L259. And the corresponding CB version during cluster creation/upgrade is stored as part of this component. Updated the patch to make use of this